### PR TITLE
generate shared dh on signup

### PR DESCRIPTION
**Pull request by mlsteele**
Post an initial shared dh key. This is very much not complete. This works for the first half (before paperkey) of signing up with the first device, and probably breaks all other cases.

Disabled by default. Enabled by being in dev mode + `--enable-shared-dh`.

### CODE REVIEW

**Comment by maxtaco** [comment]
(Rather than `int`)

File: go/engine/device_keygen.go [hunk 1]
```diff
@@ -134,36 +156,56 @@ func (e *DeviceKeygen) Push(ctx *Context, pargs *DeviceKeygenPushArgs) error {
...
 		kp, err := libkb.GenerateNaclDHKeyPair()
 		if err != nil {
 			return nil, err
 		}
 		return kp, nil
-	}, libkb.NaclDHExpireIn)
-	e.naclEncGen = libkb.NewNaclKeyGen(encArg)
+	}, e.device(), libkb.NaclDHExpireIn)
+
+	if e.args.IsEldest {
```
Comment (maxtaco) [inline go/engine/device_keygen.go]: you're doing this in all cases, even when `GetEnabledSahredDHStrict` is off?

File: go/engine/device_keygen.go [hunk 2]
```diff
@@ -191,6 +239,11 @@ func (e *DeviceKeygen) localSave(ctx *Context) {
 	if e.runErr = e.naclEncGen.SaveLKS(e.G(), e.args.Lks, ctx.LoginContext); e.runErr != nil {
 		return
 	}
+	if e.naclSharedDHGen != nil {
+		if e.runErr = e.naclSharedDHGen.SaveLKS(e.G(), e.args.Lks, ctx.LoginContext); e.runErr != nil {
```
Comment (maxtaco) [inline go/engine/device_keygen.go]: I don't want this key to his local storage, so I think we can leave this out. The plan is going to be to always sync from the server on startups.

File: go/engine/device_keygen.go [hunk 3]
```diff
@@ -242,6 +295,25 @@ func (e *DeviceKeygen) appendEncKey(ds []libkb.Delegator, ctx *Context, signer l
...
+func (e *DeviceKeygen) appendSharedDHKey(ds []libkb.Delegator, ctx *Context, signer libkb.GenericKey, eldestKID keybase1.KID, user *libkb.User) []libkb.Delegator {
+	if e.pushErr != nil {
+		return ds
+	}
+
+	e.naclSharedDHGen.UpdateArg(signer, eldestKID, libkb.DelegationTypeSharedDHKey, user)
+
+	var d libkb.Delegator
+	d, e.pushErr = e.naclSharedDHGen.Push(ctx.LoginContext, true)
+	generation := libkb.SharedDHGeneration(1)
```
Comment (maxtaco) [inline go/engine/device_keygen.go]: I called it `SharedDHKeyGeneration`

File: go/libkb/delegatekeyaggregator.go [hunk 4]
```diff
@@ -48,3 +55,37 @@ func DelegatorAggregator(lctx LoginContext, ds []Delegator) (err error) {
...
+	_, secret, err := innerKey.ExportPublicAndPrivate()
+	if err != nil {
+		return SharedDHSecretBox{}, err
+	}
+
+	encInfo, err := receiverKey.Encrypt(secret, senderKey)
+	if err != nil {
+		return SharedDHSecretBox{}, err
+	}
+	packet, err := encInfo.ToPacket()
```
Comment (maxtaco) [inline go/libkb/delegatekeyaggregator.go]: I think `PacketArmoredEncode` can do both of these in one go.

File: go/libkb/env.go [hunk 5]
```diff
@@ -650,6 +651,14 @@ func (e *Env) GetEmail() string {
 	)
 }
 
+func (e *Env) GetEnableSharedDH() bool {
+	return e.GetBool(false,
```
Comment (maxtaco) [inline go/libkb/env.go]: you'll also want to check `TestParameters` for testing.

File: go/engine/device_keygen.go [hunk 6]
```diff
@@ -108,8 +111,14 @@ func (e *DeviceKeygen) SigningKey() libkb.NaclKeyPair {
 	return e.naclSignGen.GetKeyPair()
 }
 
-func (e *DeviceKeygen) EncryptionKey() libkb.NaclKeyPair {
-	return e.naclEncGen.GetKeyPair()
+func (e DeviceKeygen) EncryptionKey() *libkb.NaclDHKeyPair {
+	k := e.naclEncGen.GetKeyPair().(libkb.NaclDHKeyPair)
+	return &k
```
Comment (mlsteele) [inline go/engine/device_keygen.go]: This just drops the data. I fixed it, but don't understand.

Comment (maxtaco) [inline go/engine/device_keygen.go]: ??? I don't understand the comment?

File: go/libkb/env.go [hunk 7]
```diff
@@ -148,8 +148,9 @@ type TestParameters struct {
 	Devel bool
 	// If we're in dev mode, the name for this test, with a random
 	// suffix.
-	DevelName  string
-	RuntimeDir string
+	DevelName      string
+	RuntimeDir     string
+	enableSharedDH bool
```
Comment (maxtaco) [inline go/libkb/env.go]: you likely need to export this to tickle it `true` in `engine`

File: go/libkb/kbsig.go [hunk 8]
```diff
@@ -32,14 +32,14 @@ func merkleRootInfo(g *GlobalContext) (ret *jsonw.Wrapper) {
...
-	IncludePGPHash     bool
-	SharedDHGeneration *SharedDHGeneration
+	Key                   GenericKey
+	EldestKID             keybase1.KID
+	ParentKID             keybase1.KID
+	HasRevSig             bool
+	RevSig                string
+	SigningUser           UserBasic
+	IncludePGPHash        bool
+	SharedDHKeyGeneration *SharedDHKeyGeneration
```
Comment (maxtaco) [inline go/libkb/kbsig.go]: Do you feel strongly about `nil` versus `0` meaning unset here?  It might make code simpler to have non-pointers. You can't have a `0` generation for shared DH keys.

Comment (mlsteele) [inline go/libkb/kbsig.go]: oh yeah, sounds good

File: go/engine/device_keygen.go (hunk 6 above)
Comment (mlsteele) [inline go/engine/device_keygen.go]: I've seen 3 versions of this function
1. `-> NaclKeyPair` on master
2. `-> *libkb.NaclDHKeyPair` some minutes ago in this branch
3. `-> libkb.NaclDHKeyPair` now
I was saying that 2 breaks everything I think by zeroing the key or something. I changed it to 3 which doesn't break any tests. I just don't understand why 2 broke that way nor really why 3 compiles.

**Comment by mlsteele** [comment]
Looks like `TestSignupEngine` works with it on. Changed `TestParameters` to a pointer so that setting it aliases. There might be other env initialization paths that need this that tests wouldn't cover?

File: go/libkb/env.go [hunk 9]
```diff
@@ -169,7 +169,7 @@ type Env struct {
 	config        ConfigReader
 	homeFinder    HomeFinder
 	writer        ConfigWriter
-	Test          TestParameters
+	Test          *TestParameters
```
Comment (maxtaco) [inline go/libkb/env.go]: why this change?

File: go/engine/signup_test.go [hunk 10]
```diff
@@ -18,10 +18,20 @@ func AssertDeviceID(g *libkb.GlobalContext) (err error) {
 }
 
 func TestSignupEngine(t *testing.T) {
+	subTestSignupEngine(t, false)
+}
+
+func TestSignupEngineSDH(t *testing.T) {
+	subTestSignupEngine(t, true)
```
Comment (maxtaco) [inline go/engine/signup_test.go]: Does this test only make an original device key, and no paper keys?

File: go/engine/device_keygen.go [hunk 11]
```diff
@@ -167,7 +167,7 @@ func (e *DeviceKeygen) Push(ctx *Context, pargs *DeviceKeygenPushArgs) error {
 }
 
 func (e *DeviceKeygen) GetEnabledSharedDHStrict() bool {
-	if e.G().Env.GetRunMode() != libkb.DevelRunMode {
+	if e.G().Env.GetRunMode() == libkb.ProductionRunMode {
```
Comment (maxtaco) [inline go/engine/device_keygen.go]: So OK for staging?

File: go/libkb/env.go (hunk 9 above)
Comment (mlsteele) [inline go/libkb/env.go]: Because otherwise [this copy](https://github.com/keybase/client/blob/master/go/libkb/test_common.go#L226) makes it so that `tc.Tp` and `tc.G.Tp` are not equal.

File: go/engine/device_keygen.go (hunk 11 above)
Comment (mlsteele) [inline go/engine/device_keygen.go]: Changed it back. I mistakenly thought test had a different run mode.

File: go/engine/signup_test.go (hunk 10 above)
Comment (mlsteele) [inline go/engine/signup_test.go]: Yeah it does [SkipPaper](https://github.com/keybase/client/blob/master/go/engine/common_test.go#L92)

**Comment by maxtaco** [comment]
OK 👍
